### PR TITLE
DR-2436: Snapshot Query

### PR DESCRIPTION
**Pull request by snf2ye**
Changes deployed here:
https://jade-sh.datarepo-dev.broadinstitute.org/

### CODE REVIEW

File: src/actions/app.js [hunk 1]
```diff
@@ -168,19 +168,30 @@ export const { hideAlert } = createActions({
...
-  [ActionTypes.RUN_QUERY]: (projectId, query, maxResults) => ({
+  [ActionTypes.RUN_QUERY]: (projectId, query) => ({
     projectId,
     query,
-    maxResults,
+  }),
+});
+
+export const { previewData } = createActions({
+  [ActionTypes.PREVIEW_DATA_SUCCESS]: (queryResults, columns) => ({
```
Comment (nmalfroy) [inline src/actions/app.js:185]: minor: could we swap the order?  I like `FOO_SUCCESS` being after `FOO`

File: src/components/table/JadeTable.jsx [hunk 2]
```diff
@@ -150,82 +106,91 @@ export class JadeTable extends React.PureComponent {
...
+                  );
+                })}
+              </TableBody>
+            )}
+          </Table>
+        )}
+        {polling && (
+          <LoadingSpinner
+            delay={delay}
+            delayMessage="For large datasets, it can take a few minutes to fetch results from BigQuery. Thank you for your patience."
```
Comment (nmalfroy) [inline src/components/table/JadeTable.jsx]: While we're in here, let's maybe take out `from BigQuery` since we're gonna want this for parquet too

File: src/reducers/query.js [hunk 3]
```diff
@@ -26,13 +32,55 @@ export default {
...
+        const rows = queryResults.map((row) => {
+          const res = {};
+          columns.forEach((col) => {
+            res[col.id] = row[col.id];
+          });
+          res.datarepo_id = row.datarepo_row_id;
+          return res;
+        });
+        const queryParams = {
+          // pageToken: queryResults.pageToken,
```
Comment (nmalfroy) [inline src/reducers/query.js]: commented code

File: src/actions/app.js (hunk 1 above)
Comment (nmalfroy) [reply src/actions/app.js:185]: should we also have a FAILURE handler?

**Review by nmalfroy — COMMENTED** [review]
Did an initial pass. looking good!  I need to click test too

File: src/actions/app.js (hunk 1 above)
Comment (snf2ye) [reply src/actions/app.js:185]: Good point - added. I added a "Failure" component as a sorta placeholder -- probably should come up with a design for handling the case when the request fails. Right now it's just an error icon + error message.

File: src/actions/app.js [hunk 4]
```diff
@@ -168,19 +168,31 @@ export const { hideAlert } = createActions({
...
-  [ActionTypes.RUN_QUERY]: (projectId, query, maxResults) => ({
+  [ActionTypes.RUN_QUERY]: (projectId, query) => ({
     projectId,
     query,
-    maxResults,
   }),
 });
 
+export const { previewData } = createActions({
+  [ActionTypes.PREVIEW_DATA]: (snapshotId, table, columns, totalRowCount) => ({
```
Comment (tlangs) [inline src/actions/app.js]: Will we be using the preview API for datasets as well as snapshots? If so, we might want to not call the id `snapshotId` here.

File: src/actions/app.js [hunk 5]
```diff
@@ -200,6 +212,18 @@ export const { applySort } = createActions({
   }),
 });
 
+export const { resetQuery } = createActions({
```
Comment (tlangs) [inline src/actions/app.js:216]: If these new actions are a part of the new preview API, shouldn't they be declared under the `previewData` constant?

File: src/components/common/query/QueryView.jsx [hunk 6]
```diff
@@ -0,0 +1,129 @@
+import React, { Fragment } from 'react';
+import PropTypes from 'prop-types';
+import { withStyles } from '@material-ui/core/styles';
+import Grid from '@material-ui/core/Grid';
+
+import { Button, Typography } from '@material-ui/core';
+
+import { Link } from 'react-router-dom';
+import SidebarDrawer from 'components/dataset/data/sidebar/SidebarDrawer';
```
Comment (tlangs) [inline src/components/common/query/QueryView.jsx]: the `SidebarDrawer` should be moved out of the `dataset` folder if its used for both datasets and snapshots.

File: src/components/common/query/QueryView.jsx [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+  sidebarWidth,
+  tableNames,
+}) {
+  return (
+    <Fragment>
+      {resourceLoaded && (
+        <div className={classes.pageRoot}>
+          <AppBreadcrumbs
+            context={{
+              type:
```
Comment (tlangs) [inline src/components/common/query/QueryView.jsx]: I would love to make typescript types for datasets and snapshots. Then, we could just pass the dataset/snapshot to `AppBreadcrumbs` and it could figure out what to do based on the type instead of needing to pass the resource type in as a parameter.

File: src/components/common/query/QueryView.jsx [hunk 8]
```diff
@@ -0,0 +1,129 @@
+import React, { Fragment } from 'react';
```
Comment (tlangs) [inline src/components/common/query/QueryView.jsx]: Since this is now referred to as the "Data" page for datasets/snapshots, can we rename files to match?

File: src/components/dataset/data/sidebar/QueryViewSidebarItem.jsx [hunk 9]
```diff
@@ -42,19 +42,19 @@
~    const { filterMap } = this.state;
~    // enable the button when there are unsaved changes
~    if (!_.isEqual(prevState.filterMap, filterMap)) {
~      this.setState({ disableButton: this.invalidChange(filterMap) });
~    }
```
Comment (tlangs) [inline src/components/dataset/data/sidebar/QueryViewSidebarItem.jsx]: Nice catch!

File: src/components/dataset/data/DatasetDataView.jsx [hunk 10]
```diff
@@ -0,0 +1,220 @@
...
+import {
+  DATAREPO_ROW_ID_COLUMN_NAME,
+  DATASET_INCLUDE_OPTIONS,
+  GOOGLE_CLOUD_RESOURCE,
+  RESOURCE_TYPE,
+} from '../../../constants';
+
+const QUERY_LIMIT = 1000;
+
+function DatasetQueryView({
```
Comment (tlangs) [inline src/components/dataset/data/DatasetDataView.jsx]: Did we decide to make all new components in Typescript? I can't remember. It would be cool to see Typescript proliferate in our codebase, though!

File: src/components/table/DatasetSnapshotsTable.jsx [hunk 11]
```diff
@@ -39,20 +39,23 @@ class DatasetSnapshotsTable extends React.PureComponent {
     const columns = [
       {
         label: 'Snapshot Name',
-        property: 'name',
+        name: 'name',
+        allowSort: true,
         render: (row) => (
-          <Link to={`/snapshots/details/${row.id}`} className={classes.jadeLink}>
+          <Link to={`/snapshots/${row.id}/overview`} className={classes.jadeLink}>
```
Comment (tlangs) [inline src/components/table/DatasetSnapshotsTable.jsx]: There's no more `overview` route. The snapshot overview pages now live at `/snapshot/{ID}`

File: src/components/table/JadeTable.jsx [hunk 12]
```diff
@@ -45,201 +36,172 @@ const styles = (theme) => ({
...
+            <JadeTableHead columns={columns} createSortHandler={createSortHandler} />
+            {!polling && (
+              <TableBody data-cy="tableBody">
+                {rows.map((row, i) => {
+                  const drId = row.datarepo_row_id;
+
+                  return (
+                    <TableRow hover tabIndex={-1} key={drId}>
+                      {columns.map((column) => {
+                        const value = handleValues(row[column.name], column);
```
Comment (tlangs) [inline src/components/table/JadeTable.jsx:141]: I like this change! Much clearer!

File: src/components/DetailViewHeader.jsx [hunk 13]
```diff
@@ -57,6 +58,10 @@ const styles = (theme) => ({
     marginTop: '20px',
     marginBottom: '10px',
   },
+  viewSnapshotButton: {
+    width: '100%',
+    marginBottom: '12px',
+  },
 });
 
 export class DetailViewHeader extends React.PureComponent {
```
Comment (tlangs) [inline src/components/DetailViewHeader.jsx]: Should this now be `OverviewHeader`?

File: src/reducers/query.js [hunk 14]
```diff
@@ -1,20 +1,28 @@
 import { handleActions } from 'redux-actions';
 import immutable from 'immutability-helper';
 import { LOCATION_CHANGE } from 'connected-react-router';
-import { ActionTypes } from 'constants/index';
+import { ActionTypes, TABLE_DEFAULT_ROWS_PER_PAGE } from 'constants/index';
 import BigQuery from 'modules/bigquery';
 
 export const queryState = {
```
Comment (tlangs) [inline src/reducers/query.js:7]: I'm wondering if its worth renaming this to `previewState`, since it's using the preview API

**Review by tlangs — CHANGES_REQUESTED** [review]
This was a hefty effort to take on. Thanks so much for chewing through it!

File: src/components/dataset/data/DatasetDataView.jsx (hunk 10 above)
Comment (snf2ye) [reply src/components/dataset/data/DatasetDataView.jsx]: Definitely - It was just more than I wanted to take on with this PR. I did attempt to rework any models to consistent var names/etc, which should help with that transition.

File: src/components/table/DatasetSnapshotsTable.jsx (hunk 11 above)
Comment (snf2ye) [reply src/components/table/DatasetSnapshotsTable.jsx]: Oh thanks, I thought I had caught all of the route changes.

File: src/components/DetailViewHeader.jsx (hunk 13 above)
Comment (snf2ye) [reply src/components/DetailViewHeader.jsx]: Yeah - I've done another pass now and renamed a bunch more files to corresponding data/overview names.

File: src/reducers/query.js (hunk 14 above)
Comment (snf2ye) [reply src/reducers/query.js:7]: There's so much tied into this "query" name - I think we can push off that name migration to when we ditch the BQ query aspect altogether. At the very least, I'd want to punt to another ticket/PR.

File: src/components/common/query/QueryView.jsx (hunk 8 above)
Comment (snf2ye) [reply src/components/common/query/QueryView.jsx]: Hopefully with my last pass, I've changed more of the components to the new names.

File: src/components/common/query/QueryView.jsx (hunk 6 above)
Comment (snf2ye) [reply src/components/common/query/QueryView.jsx]: Agreed - I think this will be covered in https://broadworkbench.atlassian.net/browse/DR-2483 - I'll add this to A.C.

File: src/actions/app.js (hunk 5 above)
Comment (snf2ye) [reply src/actions/app.js:216]: Any of the "query" actions apply to all of the tables, so it wouldn't be appropriate to move this under previewData.

File: src/components/common/data/DataView.jsx [hunk 15]
```diff
@@ -0,0 +1,129 @@
...
+  queryParams: PropTypes.object,
+  resourceId: PropTypes.string,
+  resourceLoaded: PropTypes.bool,
+  resourceName: PropTypes.string,
+  resourceType: PropTypes.string,
+  selected: PropTypes.string,
+  selectedTable: PropTypes.object,
+  sidebarWidth: PropTypes.number,
+  tableNames: PropTypes.array,
+  pageBQQuery: PropTypes.func,
```
Comment (nmalfroy) [inline src/components/common/data/DataView.jsx]: Mind just making this alphabetical (or should we get rid of this linting rule?)

File: src/components/dataset/data/DatasetDataView.jsx [hunk 16]
```diff
@@ -0,0 +1,220 @@
...
+  const [selected, setSelected] = useState('');
+  const [selectedTable, setSelectedTable] = useState(null);
+  const [sidebarWidth, setSidebarWidth] = useState(0);
+  const [canLink, setCanLink] = useState(false);
+  const [datasetLoaded, setDatasetLoaded] = useState(false);
+  const [tableNames, setTableNames] = useState([]);
+  const [panels, setPanels] = useState([]);
+
+  useEffect(() => {
+    const datasetId = match.params.uuid;
```
Comment (nmalfroy) [inline src/components/dataset/data/DatasetDataView.jsx:50]: We may want to make the datasetId be a property of the component and do the mapping in a `connect` wrapper.

File: src/components/snapshot/data/SnapshotDataView.jsx [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+function SnapshotDataView({ dispatch, match, page, queryParams, rowsPerPage, snapshot }) {
+  const [selected, setSelected] = useState('');
+  const [selectedTable, setSelectedTable] = useState(null);
+  const [sidebarWidth, setSidebarWidth] = useState(0);
+  const [snapshotLoaded, setSnapshotLoaded] = useState(false);
+  const [tableNames, setTableNames] = useState([]);
+  const [panels, setPanels] = useState([]);
+
+  useEffect(() => {
+    const snapshotId = match.params.uuid;
```
Comment (nmalfroy) [inline src/components/snapshot/data/SnapshotDataView.jsx:19]: Same comment about grabbing this from state in a connect method

**Review by nmalfroy — APPROVED** [review]
Man, this PR must have been a slog, thanks for working through it!  Is your deployed instance up to date with this code?  I'll do a list quick click test.  Let's merge this as soon as this week's deploy is done

**Review by tlangs — APPROVED** [review]
Once Muscle's comments are resolved, merge this puppy!

File: src/components/dataset/data/DatasetDataView.jsx (hunk 16 above)
Comment (snf2ye) [reply src/components/dataset/data/DatasetDataView.jsx:50]: I'm not sure how to do that. We are already pulling the match value via the connect wrapper.

There is, however,  a hooks implementation we could use instead: https://v5.reactrouter.com/web/api/Hooks/useparams

**Comment by snf2ye** [comment]
>\s your deployed instance up to date with this code? I'll do a list quick click test. Let's merge this as soon as this week's deploy is done

@nmalfroy  - Yes, latest is deployed

File: src/components/dataset/data/DatasetDataView.jsx (hunk 16 above)
Comment (snf2ye) [reply src/components/dataset/data/DatasetDataView.jsx:50]: It also seems like keeping this particular info out of the store makes sense for readability - It's clear that we're triggering updates starting from this component when the url uuid updates.

Comment (nmalfroy) [reply src/components/dataset/data/DatasetDataView.jsx:50]: I guess I was thinking of doing this mapping explicitly in the mapStateToProps method?  So at the bottom of the file you might have:

```function mapStateToProps(state) {
  return {
    datasets: state.datasets.datasets,
    datasetsCount: state.datasets.datasetsCount,
    filteredDatasetsCount: state.datasets.filteredDatasetsCount,
    features: state.user.features,
    datasetId: state.params.uuid,
  };
}``

And then datasetId is a first class prop of the component.

Basically, consolidate the mapping there...does that make sense?

**Comment by nmalfroy** [comment]
> > Is your deployed instance up to date with this code? I'll do a list quick click test. Let's merge this as soon as this week's deploy is done
> 
> @nmalfroy - Yes, latest is deployed

Cool, thanks!  I've been click testing and it looks great!

File: src/components/dataset/data/DatasetDataView.jsx (hunk 16 above)
Comment (snf2ye) [reply src/components/dataset/data/DatasetDataView.jsx:50]: Yes, that makes sense. 
I'm not sure how to populate the state.params value, though. As far as I can tell, the router state in the store doesn't currently include that information.

Comment (nmalfroy) [reply src/components/dataset/data/DatasetDataView.jsx:50]: talked offline and it's totally fine to leave
